internal/expr/scan.go

```diff
@@ -0,0 +1,196 @@
+package expr
+
+import (
+	"database/sql"
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+// ResultExpr represents the result of a database query.
+type ResultExpr struct {
+	outputs []field
+	rows    *sql.Rows
+}
+
+func NewResultExpr(pe *PreparedExpr, rows *sql.Rows) *ResultExpr {
+	return &ResultExpr{outputs: pe.outputs, rows: rows}
+}
+
+// One runs a query and decodes the first row into outputStructs.
+func (re *ResultExpr) One(outputStructs ...any) error {
+	if ok, err := re.Next(); err != nil {
+		return err
+	} else if !ok {
+		return fmt.Errorf("cannot return one row: no results")
+	}
+	err := re.Decode(outputStructs...)
+	if err != nil {
+		return err
+	}
+	re.Close()
+	return nil
+}
+
+// getTypes returns the types mentioned in expressions in the order they appear in the query.
+func getTypes(outs []field) []reflect.Type {
+	isDup := make(map[reflect.Type]bool)
+	types := []reflect.Type{}
+	for _, out := range outs {
+		if t := out.structType; !isDup[t] {
+			isDup[t] = true
+			types = append(types, t)
+		}
+	}
+	return types
+}
+
+// All iterates over the query and decodes all the rows.
+// It fabricates all struct instantiations needed.
+func (re *ResultExpr) All() ([][]any, error) {
+	var rows [][]any
+
+	outputTypes := getTypes(re.outputs)
+
+	for {
+		if ok, err := re.Next(); err != nil {
+			return [][]any{}, err
+		} else if !ok {
+			break
+		}
+
+		outputStructs := make([]reflect.Value, len(outputTypes))
+
+		for i, t := range outputTypes {
+			// reflect.New returns a pointer to a struct. It is dereferenced with Elem().
+			outputStructs[i] = reflect.New(t).Elem()
+		}
+
+		err := re.decodeRow(outputStructs...)
+		if err != nil {
+			return [][]any{}, err
+		}
+
+		row := make([]any, len(outputTypes))
+		for i, s := range outputStructs {
+			row[i] = s.Interface()
+		}
+
+		rows = append(rows, row)
+	}
+
+	re.Close()
+	return rows, nil
+}
+
+// Next prepares the next result row for reading with the Scan method.
+func (re *ResultExpr) Next() (bool, error) {
+	ok := re.rows.Next()
+	if !ok && re.rows.Err() != nil {
+		return false, re.rows.Err()
+	}
+	return ok, nil
+}
+
+// Decode copies the columns in the current row into the fields of the structs in dests.
+// dests must contain all the structs mentioned in the query.
+func (re *ResultExpr) Decode(dests ...any) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot decode expression: %s", err)
```

Comment:
s/expression/result/ - It's not the expression that is being decoded here.

Reply:
Changed